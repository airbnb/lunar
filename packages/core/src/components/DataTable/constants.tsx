import { WidthProperties } from './types';

export const SELECTION_OPTIONS = {
  ACTIVE: 'ACTIVE',
  DISABLED: 'DISABLED',
  INACTIVE: 'INACTIVE',
  HAS_ACTIVE_CHILD: 'HAS_ACTIVE_CHILD',
};

export const STATUS_OPTIONS = {
  ALERT: 'ALERT',
  WARNING: 'WARNING',
};

type HeightMap = {
  [key: string]: number;
};

export const HEIGHT_TO_PX: HeightMap = {
  micro: 32,
  small: 48,
  regular: 56,
  large: 64,
  jumbo: 108,
};

export const DEFAULT_WIDTH_PROPERTIES: WidthProperties = {
  width: 50,
  flexGrow: 1,
  flexShrink: 1,
  minWidth: 0,
};

export const EXPANDABLE_COLUMN_WIDTH = 50;

export const SELECTABLE_COLUMN_WIDTH = 50;
export const SELECTABLE_COLUMN_WIDTH_EXPANDABLE = 39;
