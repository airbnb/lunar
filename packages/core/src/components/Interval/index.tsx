import React from 'react';

// If there's no idle frame within 2 seconds, force calling
// the tick anyway, to prevent a poor user experience.
const FORCE_TICK_CALL = 2000;

export type Props = {
  /** Number of milliseconds for each interval. */
  every: number;
  /** Content to re-render. */
  children: (now: number) => NonNullable<React.ReactNode>;
};

export type State = {
  now: number;
};

/** Continuously render content at an interval. */
export default class Interval extends React.PureComponent<Props, State> {
  state = {
    now: Date.now(),
  };

  ric: number = 0;

  timer: number = 0;

  componentDidMount() {
    this.timer = window.setTimeout(this.tick, this.props.every);
  }

  componentWillUnmount() {
    // istanbul ignore next
    if (this.timer) {
      window.clearTimeout(this.timer);
    }

    // istanbul ignore next
    if (this.ric) {
      window.cancelIdleCallback(this.ric);
    }
  }

  tick = () => {
    this.timer = 0;
    this.ric = window.requestIdleCallback(
      () => {
        this.ric = 0;
        this.setState(
          {
            now: Date.now(),
          },
          () => {
            this.timer = window.setTimeout(this.tick, this.props.every);
          },
        );
      },
      {
        timeout: FORCE_TICK_CALL,
      },
    );
  };

  render() {
    return this.props.children(this.state.now);
  }
}
