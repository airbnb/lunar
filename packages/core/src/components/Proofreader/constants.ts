import { ARROW_LEFT, ARROW_UP, ARROW_DOWN, ARROW_RIGHT } from '../../keys';

export const AIRBNB_REGEX = /\b(((air|ari|iar)[bn]{3})(?!\.com))\b/gi;
export const NON_WORD_REGEX = /\W/;

export const ARROW_KEYS = [ARROW_LEFT, ARROW_UP, ARROW_DOWN, ARROW_RIGHT];

export const NO_LOCALE = 'none';
export const AUTO_DETECT_LOCALE = 'auto';
export const LOCALE_TO_LT_LOCALE: { [locale: string]: string } = {
  de: 'de-DE',
  en: 'en-US',
  ja: 'ja-JP',
  pt: 'pt-PT',
  ru: 'ru-RU',
  zh: 'zh-CN',
};
