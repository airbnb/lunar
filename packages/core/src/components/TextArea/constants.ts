// eslint-disable-next-line import/prefer-default-export
export const LANGUAGE_RULES = {
  UNNECESSARY_ADMIN_NOTE_RULE: 'AGENT_TOOLS_UNNECESSARY_ADMIN_NOTE_RULE',
};
