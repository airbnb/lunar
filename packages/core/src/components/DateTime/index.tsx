import React from 'react';
import { ToRelativeOptions } from 'luxon';
import { mutuallyExclusiveTrueProps } from 'airbnb-prop-types';
import createDateTime from '../../utils/createDateTime';
import Interval from '../Interval';
import Empty from '../Empty';
import T from '../Translate';
import toMilliseconds from '../../utils/toMilliseconds';
import { FORMAT_YMD, FORMAT_PREFIX_DAY, FORMAT_SUFFIX_TIMEZONE } from '../../constants';
import {
  timeBundle,
  dateMicroBundle,
  dateShortBundle,
  dateMediumBundle,
  dateLongBundle,
} from '../../messages';
import { DateTimeType, Locale } from '../../types';

const MINUTE_THRESHOLD = toMilliseconds('1 minute');
const MIN_RELATIVE_DATETIME_REFRESH_INTERVAL = toMilliseconds('5 seconds');
const MAX_RELATIVE_DATETIME_REFRESH_INTERVAL = toMilliseconds('6 hours');

const formatPropType = mutuallyExclusiveTrueProps(
  'clock',
  'micro',
  'short',
  'medium',
  'long',
  'relative',
  'relativeCompact',
);

export type Props = {
  /** A timestamp, either a string, number, Date object, or Luxon DateTime object. */
  at?: DateTimeType;
  /** Display the time with no date. */
  clock?: boolean;
  /** A date time pattern to format the timestamp to. */
  format?: string;
  /** Locale to translate and format the timestamp to. Defaults to "en". */
  locale?: Locale;
  /** Use long format. */
  long?: boolean;
  /** Use medium format. */
  medium?: boolean;
  /** Use micro format. */
  micro?: boolean;
  /** Remove the weekday from the display. */
  noDay?: boolean;
  /** Disable timestamps in the future. Will be reset to current timestamp. */
  noFuture?: boolean;
  /** Remove the time from the display. */
  noTime?: boolean;
  /** Remove the timezone from the display. */
  noTimezone?: boolean;
  /** Use relative format. */
  relative?: boolean;
  /** Use relative format, without 'ago' suffix (past) or 'in' prefix (future). */
  relativeCompact?: boolean;
  /** Use short format. */
  short?: boolean;
  /** A date time pattern to parse the timestamp from. */
  sourceFormat?: string;
  /** Set to a new timezone. Defaults to the client timezone or "UTC". */
  timezone?: string | boolean;
  /** Include the weekday in the display. */
  withDay?: boolean;
};

/** Display a formatted and localized timestamp. */
export default class DateTime extends React.PureComponent<Props> {
  static propTypes = {
    clock: formatPropType,
    long: formatPropType,
    medium: formatPropType,
    micro: formatPropType,
    relative: formatPropType,
    relativeCompact: formatPropType,
    short: formatPropType,
  };

  static defaultProps = {
    at: null,
    clock: false,
    format: FORMAT_YMD,
    long: false,
    medium: false,
    micro: false,
    noDay: false,
    noFuture: false,
    noTime: false,
    noTimezone: false,
    relative: false,
    relativeCompact: false,
    short: false,
    sourceFormat: '',
    withDay: false,
  };

  static format(props: Props): string {
    const {
      at,
      clock,
      format: baseFormat,
      locale,
      long,
      medium,
      micro,
      noDay,
      noFuture,
      noTime,
      noTimezone,
      relative,
      relativeCompact,
      short,
      sourceFormat,
      timezone,
      withDay,
    } = props;

    if (!at) {
      return '';
    }

    // Render a timestamp
    let timeStamp = createDateTime(at, {
      locale,
      sourceFormat,
      timezone,
    });
    let format = baseFormat || '';
    let affixDay = withDay;
    let affixTime = true;

    if (__DEV__) {
      if (!timeStamp.isValid) {
        throw new Error('Invalid timestamp passed to `DateTime`.');
      }
    }

    // Disable future dates
    if (noFuture) {
      const now = createDateTime(null, { timezone });

      if (timeStamp > now) {
        timeStamp = now;
      }
    }

    // Return early for relative timestamps
    if (relative || relativeCompact) {
      return DateTime.relative(timeStamp, { style: relativeCompact ? 'short' : 'long' });
    }

    // Determine base date format
    if (micro) {
      format = dateMicroBundle.get(locale);
    } else if (short) {
      format = dateShortBundle.get(locale);
    } else if (medium) {
      format = dateMediumBundle.get(locale);
    } else if (long) {
      format = dateLongBundle.get(locale);
      affixDay = true;
    } else if (clock) {
      format = timeBundle.get(locale);
      affixTime = false;
    } else {
      return timeStamp.toFormat(format);
    }

    // Prepend day
    if (affixDay && !noDay) {
      format = `${FORMAT_PREFIX_DAY}, ${format}`;
    }

    // Append time
    if (affixTime && !noTime) {
      format += ` ${timeBundle.get(locale)}`;
    }

    // Append timezone
    if (!noTimezone) {
      format += ` ${FORMAT_SUFFIX_TIMEZONE}`;
    }

    return timeStamp.toFormat(format);
  }

  static relative(timeStamp: DateTimeType, options: ToRelativeOptions = {}): string {
    const relative = createDateTime(timeStamp);
    const diff = DateTime.diff(relative, options.base);
    const fewPhrase =
      options.style === 'narrow'
        ? T.phrase(
            'a few sec.',
            {},
            { context: 'Relative time within a minute', key: 'lunar.datetime.secsAgoNarrow' },
          )
        : T.phrase(
            'a few seconds',
            {},
            { context: 'Relative time within a minute', key: 'lunar.datetime.secsAgo' },
          );

    if (diff > 0 && diff < MINUTE_THRESHOLD) {
      if (!options.style || options.style === 'long') {
        return T.phrase(
          'in %{time}',
          { time: fewPhrase },
          {
            context: 'Relative time explaining something will happen soon',
            key: 'lunar.datetime.in',
          },
        );
      }

      return fewPhrase;
    }

    if (diff <= 0 && diff > -MINUTE_THRESHOLD) {
      if (!options.style || options.style === 'long') {
        return T.phrase(
          '%{time} ago',
          { time: fewPhrase },
          {
            context: 'Relative time explaining something recently happened',
            key: 'lunar.datetime.ago',
          },
        );
      }

      return fewPhrase;
    }

    return relative.toRelative(options) || relative.toFormat(FORMAT_YMD);
  }

  static diff(to: DateTimeType, from: DateTimeType | null = null): number {
    return (
      createDateTime(to, { timezone: 'UTC' }).toMillis() -
      createDateTime(from, { timezone: 'UTC' }).toMillis()
    );
  }

  getRefreshInterval() {
    const { at, sourceFormat } = this.props;
    const difference = Math.abs(DateTime.diff(createDateTime(at, { sourceFormat })));

    // Decay refresh rate based on how long its been since the given timestamp
    // < 1 minute: update every 5 seconds
    // 10 minutes: update every 1 minute
    // 1 hour: update every 6 minutes
    // 1 day: update every 2.4 hours
    // > 2 day: update every 6 hours
    return Math.min(
      Math.max(difference / 10, MIN_RELATIVE_DATETIME_REFRESH_INTERVAL),
      MAX_RELATIVE_DATETIME_REFRESH_INTERVAL,
    );
  }

  rfc() {
    const { at, sourceFormat } = this.props;

    return createDateTime(at, { sourceFormat }).toFormat("yyyy-MM-dd'T'HH:mm:ssZZ"); // RFC3339
  }

  renderTimeElement = () => {
    const formatted = DateTime.format(this.props);

    if (!formatted) {
      return <Empty />;
    }

    return <time dateTime={this.rfc()}>{formatted}</time>;
  };

  render() {
    if (!this.props.relative) {
      return this.renderTimeElement();
    }

    return (
      <Interval key={this.rfc()} every={this.getRefreshInterval()}>
        {this.renderTimeElement}
      </Interval>
    );
  }
}
