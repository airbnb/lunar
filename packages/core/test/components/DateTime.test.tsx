import React from 'react';
import { shallow } from 'enzyme';
import { DateTime as LuxonDateTime } from 'luxon';
import DateTime from '../../src/components/DateTime';
import Empty from '../../src/components/Empty';
import { FORMAT_YMD, LOCALES } from '../../src/constants';
import createDateTime from '../../src/utils/createDateTime';
import getLanguageFromLocale from '../../src/utils/getLanguageFromLocale';
import toMilliseconds from '../../src/utils/toMilliseconds';
import Interval from '../../src/components/Interval';

describe('<DateTime />', () => {
  let date: Date;

  beforeEach(() => {
    date = new Date(Date.UTC(1988, 1, 26, 16, 12, 5));
  });

  it('renders empty if no time', () => {
    const wrapper = shallow(<DateTime timezone="UTC" />);

    expect(wrapper.find(Empty)).toHaveLength(1);
  });

  it('renders a <time> element', () => {
    const wrapper = shallow(<DateTime at={date} timezone="UTC" />);

    expect(wrapper.find('time')).toHaveLength(1);
  });

  it('errors for an invalid time', () => {
    expect(() => shallow(<DateTime at="2016-02-33" timezone="UTC" />)).toThrow(
      'Invalid DateTime: unit out of range: you specified 33 (of type number) as a day, which is invalid',
    );
  });

  it('accepts a UNIX timestamp in milliseconds', () => {
    const wrapper = shallow(<DateTime medium at={date.getTime()} timezone="UTC" />);

    expect(wrapper.text()).toBe('Feb 26, 1988 4:12PM UTC');
  });

  it('accepts a custom format', () => {
    const wrapper = shallow(
      <DateTime
        medium
        at="26/02/1988 - 16:12:05"
        timezone="UTC"
        sourceFormat="dd/MM/yyyy - HH:mm:ss"
      />,
    );

    expect(wrapper.text()).toBe('Feb 26, 1988 4:12PM UTC');
  });

  it('disables the timezone', () => {
    const rfc3339 = "yyyy-MM-dd'T'HH:mm:ssZZ";
    const wrapper = shallow(<DateTime at={date} timezone={false} format={rfc3339} />);
    const wrapperUTC = shallow(
      <DateTime at={date} timezone="America/Los_Angeles" format={rfc3339} />,
    );

    expect(wrapper.text()).toBe('1988-02-26T16:12:05+00:00');
    expect(wrapperUTC.text()).toBe('1988-02-26T08:12:05-08:00');
  });

  it('formats to FORMAT_YMD', () => {
    const wrapper = shallow(<DateTime at={date} timezone="UTC" format={FORMAT_YMD} />);

    expect(wrapper.text()).toBe('1988-02-26');
  });

  it('formats using the `micro` prop', () => {
    const wrapper = shallow(<DateTime micro at={date} timezone="UTC" />);

    expect(wrapper.text()).toBe('Feb 26 4:12PM UTC');
  });

  it('formats using the `short` prop', () => {
    const wrapper = shallow(<DateTime short at={date} timezone="UTC" />);

    expect(wrapper.text()).toBe('Feb 26, 88 4:12PM UTC');
  });

  it('formats using the `medium` prop', () => {
    const wrapper = shallow(<DateTime medium at={date} timezone="UTC" />);

    expect(wrapper.text()).toBe('Feb 26, 1988 4:12PM UTC');
  });

  it('formats using the `long` prop', () => {
    const wrapper = shallow(<DateTime long at={date} timezone="UTC" />);

    expect(wrapper.text()).toBe('Fri, February 26, 1988 4:12PM UTC');
  });

  it('formats using the `clock` prop', () => {
    const wrapper = shallow(<DateTime clock at={date} timezone="UTC" />);

    expect(wrapper.text()).toBe('4:12PM UTC');
  });

  it('formats without the timestamp', () => {
    const wrapper = shallow(<DateTime long noTime at={date} timezone="UTC" />);

    expect(wrapper.text()).toBe('Fri, February 26, 1988 UTC');
  });

  it('formats with the timestamp but doesnt override format', () => {
    const wrapper = shallow(<DateTime at={date} timezone="UTC" format="yyyy-MM-dd" />);

    expect(wrapper.text()).toBe('1988-02-26');
  });

  it('displays the day', () => {
    const micro = shallow(<DateTime micro withDay at={date} timezone="UTC" />);
    const short = shallow(<DateTime short withDay at={date} timezone="UTC" />);
    const medium = shallow(<DateTime medium withDay at={date} timezone="UTC" />);
    const long = shallow(<DateTime long at={date} timezone="UTC" />); // On by default

    expect(micro.text()).toBe('Fri, Feb 26 4:12PM UTC');
    expect(short.text()).toBe('Fri, Feb 26, 88 4:12PM UTC');
    expect(medium.text()).toBe('Fri, Feb 26, 1988 4:12PM UTC');
    expect(long.text()).toBe('Fri, February 26, 1988 4:12PM UTC');
  });

  it('hides the day', () => {
    const micro = shallow(<DateTime micro at={date} timezone="UTC" />);
    const short = shallow(<DateTime short at={date} timezone="UTC" />);
    const medium = shallow(<DateTime medium at={date} timezone="UTC" />);
    const long = shallow(<DateTime long noDay at={date} timezone="UTC" />); // On by default

    expect(micro.text()).toBe('Feb 26 4:12PM UTC');
    expect(short.text()).toBe('Feb 26, 88 4:12PM UTC');
    expect(medium.text()).toBe('Feb 26, 1988 4:12PM UTC');
    expect(long.text()).toBe('February 26, 1988 4:12PM UTC');
  });

  it('hides the time', () => {
    const micro = shallow(<DateTime micro noTime at={date} timezone="UTC" />);
    const short = shallow(<DateTime short noTime at={date} timezone="UTC" />);
    const medium = shallow(<DateTime medium noTime at={date} timezone="UTC" />);
    const long = shallow(<DateTime long noTime at={date} timezone="UTC" />);

    expect(micro.text()).toBe('Feb 26 UTC');
    expect(short.text()).toBe('Feb 26, 88 UTC');
    expect(medium.text()).toBe('Feb 26, 1988 UTC');
    expect(long.text()).toBe('Fri, February 26, 1988 UTC');
  });

  it('hides the timezone', () => {
    const micro = shallow(<DateTime micro noTimezone at={date} timezone="UTC" />);
    const short = shallow(<DateTime short noTimezone at={date} timezone="UTC" />);
    const medium = shallow(<DateTime medium noTimezone at={date} timezone="UTC" />);
    const long = shallow(<DateTime long noTimezone at={date} timezone="UTC" />);

    expect(micro.text()).toBe('Feb 26 4:12PM');
    expect(short.text()).toBe('Feb 26, 88 4:12PM');
    expect(medium.text()).toBe('Feb 26, 1988 4:12PM');
    expect(long.text()).toBe('Fri, February 26, 1988 4:12PM');
  });

  it('hides all the things', () => {
    const micro = shallow(<DateTime micro noTime noTimezone noDay at={date} timezone="UTC" />);
    const short = shallow(<DateTime short noTime noTimezone noDay at={date} timezone="UTC" />);
    const medium = shallow(<DateTime medium noTime noTimezone noDay at={date} timezone="UTC" />);
    const long = shallow(<DateTime long noTime noTimezone noDay at={date} timezone="UTC" />);

    expect(micro.text()).toBe('Feb 26');
    expect(short.text()).toBe('Feb 26, 88');
    expect(medium.text()).toBe('Feb 26, 1988');
    expect(long.text()).toBe('February 26, 1988');
  });

  describe('localized messages', () => {
    LOCALES.forEach(locale => {
      it(`${getLanguageFromLocale(locale)} - ${locale}`, () => {
        const props = {
          at: date,
          timezone: 'UTC',
          noTime: true,
          noTimezone: true,
        };
        const micro = shallow(<DateTime {...props} micro locale={locale} />);
        const short = shallow(<DateTime {...props} short locale={locale} />);
        const medium = shallow(<DateTime {...props} medium locale={locale} />);
        const long = shallow(<DateTime {...props} long locale={locale} />);
        const clock = shallow(<DateTime {...props} clock locale={locale} />);

        expect(micro.text()).toMatchSnapshot();
        expect(short.text()).toMatchSnapshot();
        expect(medium.text()).toMatchSnapshot();
        expect(long.text()).toMatchSnapshot();
        expect(clock.text()).toMatchSnapshot();
      });
    });
  });

  describe('relative time', () => {
    describe('intervals', () => {
      it('renders an Interval component when relative is set', () => {
        const wrapper = shallow(<DateTime relative at={date} timezone="UTC" />);

        expect(wrapper.find(Interval)).toHaveLength(1);
      });

      it('does not render an Interval component when relativeCompact is set', () => {
        const wrapper = shallow(<DateTime relativeCompact at={date} timezone="UTC" />);

        expect(wrapper.find(Interval)).toHaveLength(0);
      });

      it('uses 5 seconds as the minimal Interval update', () => {
        const wrapper = shallow(<DateTime relative timezone="UTC" />);

        expect(wrapper.find(Interval).props().every).toBe(toMilliseconds('5 seconds'));
      });

      it('uses 6 hours as the maximum Interval update', () => {
        date.setFullYear(2088);
        const wrapper = shallow(<DateTime relative at={date} timezone="UTC" />);

        expect(wrapper.find(Interval).props().every).toBe(toMilliseconds('6 hours'));
      });
    });

    describe('rendered output', () => {
      it('formats to relative time in the future', () => {
        const wrapper = shallow(
          <DateTime relative at={createDateTime().plus({ years: 70 })} timezone="UTC" />,
        ).dive();

        expect(wrapper.text()).toBe('in 69 years');
      });

      it('uses the current time when noFuture is set and the time is in the future', () => {
        date.setFullYear(2088);

        const wrapper = shallow(<DateTime relative noFuture at={date} timezone="UTC" />).dive();

        expect(wrapper.text()).toBe('a few seconds ago');
      });

      it('uses the provided time when noFuture is set and the time is in the past', () => {
        const wrapper = shallow(
          <DateTime relative noFuture at={createDateTime().minus({ seconds: 5 })} timezone="UTC" />,
        ).dive();

        expect(wrapper.text()).toBe('a few seconds ago');
      });
    });
  });

  describe('relative()', () => {
    let base: LuxonDateTime;

    beforeEach(() => {
      base = createDateTime(new Date(Date.UTC(2019, 1, 1)));
    });

    describe('past times', () => {
      it('handles seconds', () => {
        expect(DateTime.relative(base.minus({ seconds: 44 }), { base })).toBe('a few seconds ago');
      });

      it('handles minutes', () => {
        expect(DateTime.relative(base.minus({ minutes: 1 }), { base })).toBe('1 minute ago');
      });

      it('handles hours', () => {
        expect(DateTime.relative(base.minus({ hours: 1 }), { base })).toBe('1 hour ago');
      });

      it('handles days', () => {
        expect(DateTime.relative(base.minus({ days: 1 }), { base })).toBe('1 day ago');
      });

      it('handles months', () => {
        expect(DateTime.relative(base.minus({ months: 1 }), { base })).toBe('1 month ago');
      });

      it('handles years', () => {
        expect(DateTime.relative(base.minus({ years: 1 }), { base })).toBe('1 year ago');
        expect(DateTime.relative(base.minus({ years: 2 }), { base })).toBe('2 years ago');
      });
    });

    describe('future times', () => {
      it('handles seconds', () => {
        expect(DateTime.relative(base.plus({ seconds: 44 }), { base })).toBe('in a few seconds');
      });

      it('handles minutes', () => {
        expect(DateTime.relative(base.plus({ minutes: 1.1 }), { base })).toBe('in 1 minute');
      });

      it('handles hours', () => {
        expect(DateTime.relative(base.plus({ minutes: 70 }), { base })).toBe('in 1 hour');
      });

      it('handles days', () => {
        expect(DateTime.relative(base.plus({ hours: 25 }), { base })).toBe('in 1 day');
      });

      it('handles months', () => {
        expect(DateTime.relative(base.plus({ days: 32 }), { base })).toBe('in 1 month');
      });

      it('handles years', () => {
        expect(DateTime.relative(base.plus({ months: 13 }), { base })).toBe('in 1 year');
        expect(DateTime.relative(base.plus({ months: 25 }), { base })).toBe('in 2 years');
      });
    });

    describe('compact relative times', () => {
      describe('past times', () => {
        it('handles seconds', () => {
          expect(DateTime.relative(base.minus({ seconds: 44 }), { base, style: 'short' })).toBe(
            'a few seconds',
          );
        });

        it('handles minutes', () => {
          expect(DateTime.relative(base.minus({ minutes: 1 }), { base, style: 'short' })).toBe(
            '1 min. ago',
          );
        });

        it('handles hours', () => {
          expect(DateTime.relative(base.minus({ hours: 1 }), { base, style: 'short' })).toBe(
            '1 hr. ago',
          );
        });

        it('handles days', () => {
          expect(DateTime.relative(base.minus({ hours: 24 }), { base, style: 'short' })).toBe(
            '1 day ago',
          );
        });

        it('handles months', () => {
          expect(DateTime.relative(base.minus({ days: 31 }), { base, style: 'short' })).toBe(
            '1 mo. ago',
          );
        });

        it('handles years', () => {
          expect(DateTime.relative(base.minus({ years: 1 }), { base, style: 'short' })).toBe(
            '1 yr. ago',
          );
          expect(DateTime.relative(base.minus({ months: 24 }), { base, style: 'short' })).toBe(
            '2 yr. ago',
          );
        });
      });

      describe('future times', () => {
        it('handles seconds', () => {
          expect(DateTime.relative(base.plus({ seconds: 44 }), { base, style: 'narrow' })).toBe(
            'a few sec.',
          );
        });

        it('handles minutes', () => {
          expect(DateTime.relative(base.plus({ minutes: 1.1 }), { base, style: 'narrow' })).toBe(
            'in 1 min.',
          );
        });

        it('handles hours', () => {
          expect(DateTime.relative(base.plus({ hours: 1.1 }), { base, style: 'narrow' })).toBe(
            'in 1 hr.',
          );
        });

        it('handles days', () => {
          expect(DateTime.relative(base.plus({ hours: 25 }), { base, style: 'narrow' })).toBe(
            'in 1 day',
          );
        });

        it('handles months', () => {
          expect(DateTime.relative(base.plus({ days: 32 }), { base, style: 'narrow' })).toBe(
            'in 1 mo.',
          );
        });

        it('handles years', () => {
          expect(DateTime.relative(base.plus({ months: 13 }), { base, style: 'narrow' })).toBe(
            'in 1 yr.',
          );
          expect(DateTime.relative(base.plus({ months: 26 }), { base, style: 'narrow' })).toBe(
            'in 2 yr.',
          );
        });
      });
    });

    it('handles no rounding logic', () => {
      expect(DateTime.relative(base.minus({ seconds: 44 }), { base, round: false })).toBe(
        'a few seconds ago',
      );

      expect(
        DateTime.relative(base.minus({ minutes: 1, seconds: 31 }), { base, round: false }),
      ).toBe('1.51 minutes ago');

      expect(DateTime.relative(base.minus({ minutes: 44 }), { base, round: false })).toBe(
        '44 minutes ago',
      );

      expect(DateTime.relative(base.minus({ minutes: 45 }), { base, round: false })).toBe(
        '45 minutes ago',
      );

      expect(DateTime.relative(base.minus({ hours: 1, minutes: 31 }), { base, round: false })).toBe(
        '1.51 hours ago',
      );

      expect(DateTime.relative(base.minus({ hours: 22 }), { base, round: false })).toBe(
        '22 hours ago',
      );

      expect(DateTime.relative(base.minus({ hours: 33 }), { base, round: false })).toBe(
        '1.37 days ago',
      );

      expect(DateTime.relative(base.minus({ days: 25 }), { base, round: false })).toBe(
        '25 days ago',
      );

      expect(DateTime.relative(base.minus({ days: 47 }), { base, round: false })).toBe(
        '1.51 months ago',
      );

      expect(DateTime.relative(base.minus({ months: 10 }), { base, round: false })).toBe(
        '10 months ago',
      );

      expect(DateTime.relative(base.minus({ months: 21 }), { base, round: false })).toBe(
        '1.75 years ago',
      );
    });
  });
});
