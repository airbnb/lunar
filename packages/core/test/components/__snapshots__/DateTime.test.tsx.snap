// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DateTime /> localized messages Dutch - nl 1`] = `"26 feb."`;

exports[`<DateTime /> localized messages Dutch - nl 2`] = `"26 feb. 88"`;

exports[`<DateTime /> localized messages Dutch - nl 3`] = `"26 feb. 1988"`;

exports[`<DateTime /> localized messages Dutch - nl 4`] = `"vr, 26 februari 1988"`;

exports[`<DateTime /> localized messages Dutch - nl 5`] = `"16:12"`;

exports[`<DateTime /> localized messages English - en 1`] = `"Feb 26"`;

exports[`<DateTime /> localized messages English - en 2`] = `"Feb 26, 88"`;

exports[`<DateTime /> localized messages English - en 3`] = `"Feb 26, 1988"`;

exports[`<DateTime /> localized messages English - en 4`] = `"Fri, February 26, 1988"`;

exports[`<DateTime /> localized messages English - en 5`] = `"4:12PM"`;

exports[`<DateTime /> localized messages French - fr 1`] = `"26 févr."`;

exports[`<DateTime /> localized messages French - fr 2`] = `"26 févr. 88"`;

exports[`<DateTime /> localized messages French - fr 3`] = `"26 févr. 1988"`;

exports[`<DateTime /> localized messages French - fr 4`] = `"ven., 26 février 1988"`;

exports[`<DateTime /> localized messages French - fr 5`] = `"16:12"`;

exports[`<DateTime /> localized messages German - de 1`] = `"26. Feb"`;

exports[`<DateTime /> localized messages German - de 2`] = `"26. Feb 88"`;

exports[`<DateTime /> localized messages German - de 3`] = `"26. Feb 1988"`;

exports[`<DateTime /> localized messages German - de 4`] = `"Fr, 26. Februar 1988"`;

exports[`<DateTime /> localized messages German - de 5`] = `"16:12"`;

exports[`<DateTime /> localized messages Italian - it 1`] = `"26 feb"`;

exports[`<DateTime /> localized messages Italian - it 2`] = `"26 feb 88"`;

exports[`<DateTime /> localized messages Italian - it 3`] = `"26 feb 1988"`;

exports[`<DateTime /> localized messages Italian - it 4`] = `"ven, 26 febbraio 1988"`;

exports[`<DateTime /> localized messages Italian - it 5`] = `"16:12"`;

exports[`<DateTime /> localized messages Japanese - ja 1`] = `"2月26日"`;

exports[`<DateTime /> localized messages Japanese - ja 2`] = `"88年2月26日"`;

exports[`<DateTime /> localized messages Japanese - ja 3`] = `"1988年2月26日"`;

exports[`<DateTime /> localized messages Japanese - ja 4`] = `"金, 1988年2月26日"`;

exports[`<DateTime /> localized messages Japanese - ja 5`] = `"16:12"`;

exports[`<DateTime /> localized messages Korean - ko 1`] = `"2월 26일"`;

exports[`<DateTime /> localized messages Korean - ko 2`] = `"88년 2월 26일"`;

exports[`<DateTime /> localized messages Korean - ko 3`] = `"1988년 2월 26일"`;

exports[`<DateTime /> localized messages Korean - ko 4`] = `"금, 1988년 2월 26일"`;

exports[`<DateTime /> localized messages Korean - ko 5`] = `"오후 4:12"`;

exports[`<DateTime /> localized messages Malay - ms 1`] = `"26 Feb"`;

exports[`<DateTime /> localized messages Malay - ms 2`] = `"26 Feb 88"`;

exports[`<DateTime /> localized messages Malay - ms 3`] = `"26 Feb 1988"`;

exports[`<DateTime /> localized messages Malay - ms 4`] = `"Jum, 26 Februari 1988"`;

exports[`<DateTime /> localized messages Malay - ms 5`] = `"16:12"`;

exports[`<DateTime /> localized messages Mandarin - zh 1`] = `"2月26日"`;

exports[`<DateTime /> localized messages Mandarin - zh 2`] = `"88年2月26日"`;

exports[`<DateTime /> localized messages Mandarin - zh 3`] = `"1988年2月26日"`;

exports[`<DateTime /> localized messages Mandarin - zh 4`] = `"周五, 1988年二月26日"`;

exports[`<DateTime /> localized messages Mandarin - zh 5`] = `"16:12"`;

exports[`<DateTime /> localized messages Portuguese - pt 1`] = `"26 de fev"`;

exports[`<DateTime /> localized messages Portuguese - pt 2`] = `"26 de fev de 88"`;

exports[`<DateTime /> localized messages Portuguese - pt 3`] = `"26 de fev de 1988"`;

exports[`<DateTime /> localized messages Portuguese - pt 4`] = `"sex, 26 de fevereiro de 1988"`;

exports[`<DateTime /> localized messages Portuguese - pt 5`] = `"16:12"`;

exports[`<DateTime /> localized messages Russian - ru 1`] = `"26 февр. г."`;

exports[`<DateTime /> localized messages Russian - ru 2`] = `"26 февр. 88 г."`;

exports[`<DateTime /> localized messages Russian - ru 3`] = `"26 февр. 1988 г."`;

exports[`<DateTime /> localized messages Russian - ru 4`] = `"пт, 26 февраль 1988 г."`;

exports[`<DateTime /> localized messages Russian - ru 5`] = `"16:12"`;

exports[`<DateTime /> localized messages Spanish - es 1`] = `"26 de feb."`;

exports[`<DateTime /> localized messages Spanish - es 2`] = `"26 de feb. de 88"`;

exports[`<DateTime /> localized messages Spanish - es 3`] = `"26 de feb. de 1988"`;

exports[`<DateTime /> localized messages Spanish - es 4`] = `"vie., 26 de febrero de 1988"`;

exports[`<DateTime /> localized messages Spanish - es 5`] = `"16:12"`;

exports[`<DateTime /> localized messages Turkish - tr 1`] = `"26 Şub"`;

exports[`<DateTime /> localized messages Turkish - tr 2`] = `"26 Şub 88"`;

exports[`<DateTime /> localized messages Turkish - tr 3`] = `"26 Şub 1988"`;

exports[`<DateTime /> localized messages Turkish - tr 4`] = `"Cum, 26 Şubat 1988"`;

exports[`<DateTime /> localized messages Turkish - tr 5`] = `"16:12"`;
